verilog/tools/ls/autoexpand.cc

```diff
@@ -89,33 +154,156 @@ class AutoExpander {
   // Token context created from the buffer contents
   const TokenInfo::Context context_;
 
+  // Symbol table wrapper for the language server
+  SymbolTableHandler &symbol_table_handler;
+
   // String to add at the end of each generated line
-  std::string line_end_str_;
+  std::string indent_;
 
   // Regex for finding AUTOARG comments
-  static std::regex autoarg_re_;
+  static const std::regex autoarg_re_;
+
+  // Regex for finding AUTOINST comments
+  static const std::regex autoinst_re_;
+
+  // Regexes for AUTO_TEMPLATE comments
+  static const std::regex autotemplate_detect_re_;
+  static const std::regex autotemplate_type_re_;
+  static const std::regex autotemplate_conn_re_;
 };
 
-std::regex AutoExpander::autoarg_re_{R"(/\*\s*AUTOARG\s*\*/)"};
+const std::regex AutoExpander::autoarg_re_{R"(/\*\s*AUTOARG\s*\*/)"};
+
+const std::regex AutoExpander::autoinst_re_{R"(/\*\s*AUTOINST\s*\*/)"};
 
-void AutoExpander::EmitPortDeclarations(
+// AUTO_TEMPLATE regex breakdown:
+// /\*                             – start of comment
+// (\s*\S+\s+AUTO_TEMPLATE\s*\n)*  – optional other AUTO_TEMPLATE types, end
+//                                   with newline
+// \s*\S+\s+AUTO_TEMPLATE          – at least one AUTO_TEMPLATE is required
+// \s*(".*")?                      – optional instance name regex
+// \s*\([\s\S]*?\);                – parens with port connections
+// \s*\*/                          – end of comment
+const std::regex AutoExpander::autotemplate_detect_re_{
+    R"(/\*(\s*\S+\s+AUTO_TEMPLATE\s*\n)*\s*\S+\s+AUTO_TEMPLATE\s*(".*")?\s*\([\s\S]*?\);\s*\*/)"};
+
+// AUTO_TEMPLATE type regex: the first capturing group is the instance type
+const std::regex AutoExpander::autotemplate_type_re_{
+    R"((\S+)\s+AUTO_TEMPLATE)"};
+
+// AUTO_TEMPLATE connection regex breakdown:
+// \.\s*      – starts with a dot
+// ([^\s(]+?) – first group, at least one character other than whitespace or
+//              opening paren
+// \s*\(\s*   – optional whitespace, opening paren, optional whitespace again
+// ([^\s(]+?) – second group, same as the first one
+// \s*\)*     – optional whitespace, closing paren
+const std::regex AutoExpander::autotemplate_conn_re_{
+    R"(\.\s*([^\s(]+?)\s*\(\s*([^\s(]+?)\s*\))"};
+
+const AutoExpander::Template *AutoExpander::Template::from(
+    const Map &templates, const absl::string_view type_id,
+    const long int instance_offset) {
```

Comment:
(same comment as above: instead of using long int, use `int64_t` from `<cstdint>` to emphasize that you want to potentially use large values (`long int` is defined differently on different platforms).